homeassistant/components/nws/weather.py

```diff
@@ -140,96 +139,72 @@ def __init__(
         self.nws = nws_data.api
         latitude = entry_data[CONF_LATITUDE]
         longitude = entry_data[CONF_LONGITUDE]
-        self.coordinator_forecast_legacy = nws_data.coordinator_forecast
-        self.station = self.nws.station
 
-        self.observation: dict[str, Any] | None = None
-        self._forecast_hourly: list[dict[str, Any]] | None = None
-        self._forecast_legacy: list[dict[str, Any]] | None = None
-        self._forecast_twice_daily: list[dict[str, Any]] | None = None
+        self.station = self.nws.station
 
         self._attr_unique_id = _calculate_unique_id(entry_data, DAYNIGHT)
         self._attr_device_info = device_info(latitude, longitude)
         self._attr_name = self.station
 
     async def async_added_to_hass(self) -> None:
-        """Set up a listener and load data."""
+        """When entity is added to hass."""
         await super().async_added_to_hass()
-        self.async_on_remove(
-            self.coordinator_forecast_legacy.async_add_listener(
-                self._handle_legacy_forecast_coordinator_update
-            )
-        )
-        # Load initial data from coordinators
-        self._handle_coordinator_update()
-        self._handle_hourly_forecast_coordinator_update()
-        self._handle_twice_daily_forecast_coordinator_update()
-        self._handle_legacy_forecast_coordinator_update()
-
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Load data from integration."""
-        self.observation = self.nws.observation
-        self.async_write_ha_state()
-
-    @callback
-    def _handle_hourly_forecast_coordinator_update(self) -> None:
-        """Handle updated data from the hourly forecast coordinator."""
-        self._forecast_hourly = self.nws.forecast_hourly
+        self.async_on_remove(partial(self._remove_forecast_listener, "daily"))
+        self.async_on_remove(partial(self._remove_forecast_listener, "hourly"))
+        self.async_on_remove(partial(self._remove_forecast_listener, "twice_daily"))
 
-    @callback
-    def _handle_twice_daily_forecast_coordinator_update(self) -> None:
-        """Handle updated data from the twice daily forecast coordinator."""
-        self._forecast_twice_daily = self.nws.forecast
+        if self.forecast_coordinators["twice_daily"] is not None:
```

Comment:
I'd probably write that as

```python
        for forecast_type in ("twice_daily", "hourly"):
            if (coordinator := self.forecast_coordinators[forecast_type]) is None:
                continue
            self.unsub_forecast[forecast_type] = coordinator.async_add_listener(
                partial(self._handle_forecast_update, forecast_type)
            )
```

Reply:
Thanks! An even better style IMO would be to import the `"twice_daily"` and `"hourly"` strings from `homeassistant.components.weather`.  Can that be done in this PR, or should that be done in a separate PR?

```py
        for forecast_type in (TWICE_DAILY, HOURLY):
            if (coordinator := self.forecast_coordinators[forecast_type]) is None:
                continue
            self.unsub_forecast[forecast_type] = coordinator.async_add_listener(
                partial(self._handle_forecast_update, forecast_type)
            )
```

Reply:
> what I mean is that `homeassistant.components.weather` does not have these constants defined, should we add them into that integration?]

Please don't change core base platforms in an integration PR. It should be a separate PR

Reply:
Addressed in https://github.com/home-assistant/core/pull/115857/commits/e6c33093d5c7eb1027fedde64c71490636308034.  I avoided the temptation.  I can make a separate PR once this one is merged to avoid conflicts.